### DISCUSSION TITLE
How to insert(replace) text with 90° rotation

### DISCUSSION THREAD

**Discussion by kuaidianbiye1994** [pymupdf/PyMuPDF#1750] [Looking for help]
Hello,
I want to make a train timetable translate programm for my master’s thesis, and PyMuPDF reaches most of my goal. But still I have some needs that I can't manage on my own..

For example, I have a german train timetable to be translated to a chinese train timetable:
![deutsche Fahrplan ](https://user-images.githubusercontent.com/103218123/173199318-9b010f8f-c0aa-493d-8f57-055d8b9a0257.png)
For translating this into a chinese one, I have to rotate this timetable 90 degrees, like this:
![chinese fahrplan](https://user-images.githubusercontent.com/103218123/173199531-b805ed0c-2e49-4d9f-b9aa-ebab65762e6b.png)
Now I have a problem, the number on the x-axis(like 4,5 on the picture, which refers to time) ,they "lie" on the x-axis, but I want them to "stand" on x-axis. In other words, I want to rotate these numbers 90 degrees.
The same problem is, the text translated (like" Leipzig Hbf"), I managed to translate them into chinese but I also need a 90-degree-rotation.

I used 
`
 def set_font_de(doc, font_name, font_file):

    doc[0].insert_text((0, 0), " ", fontname=font_name,fontfile=font_file, render_mode=3,rotate=90)  #to make space for new text
`
 
I also tried 
` 
 for page in doc:

        spans = collect_span_de(page)
        for span in spans:
            text = translate_span_de(translator, span, to_lang)
            print(f'translated {span["text"]} to {text}...')
            replace_span(page, span, text, font_name)
            page.setRotation(90)    # I tried to rotate the page before/after I replace the text, but two ways have no difference
`
Unluckily, these methods don't work well : (

Thanks for any help in advance!
[Deutsche Fahrplan 1.pdf](https://github.com/pymupdf/PyMuPDF/files/8884490/Deutsche.Fahrplan.1.pdf)

**Comment by kuaidianbiye1994** [comment]
I'm still looking for any possible help. Thanks for any help in advance!

**Comment by JorjMcKie** [comment]
Use `page.insert_text()` or `page.insert_textbox()`. Both support a `rotate` parameter.

Based on `rotate`, the textbox method will automatically determine the correct corner of the rectangle, where writing the text should begin. So with `rotate=90` it will be the bottom-left corner and the lines will fill vertically from left to right.
Why don't you look at the documentation [here](https://pymupdf.readthedocs.io/en/latest/shape.html#Shape.insert_textbox).

As a general comment:
You cannot simply "replace" text on an existing PDF page. You must either use an empty page, or you must remove existing text first, e.g. by using redaction annotations.

If you use render mode 3, the new text will be invisible, BTW.
